File: src/api/practice.ts
import { api } from './client';

export interface DailyStats {
  date: string;
  avg_cpm: number;
  avg_accuracy: number;
  session_count: number;
}

export interface StatsResponse {
  daily: DailyStats[];
  best_cpm: number;
  total_sessions: number;
}

export interface SessionCreate {
  mode: string;
  lang: string;
  cpm: number;
  accuracy: number;
  duration: number;
  sentences?: number;
  error_log?: Record<string, number>;
}

export const practiceApi = {
  getStats:      (days = 30) => api.get<StatsResponse>(`/practice/stats?days=${days}`),
  createSession: (body: SessionCreate) => api.post<void>('/practice/sessions', body),
};

File: backend/app/schemas/practice.py
from datetime import datetime
from pydantic import BaseModel


class SessionCreate(BaseModel):
    mode: str
    lang: str
    cpm: int
    accuracy: int
    duration: int
    sentences: int = 0
    error_log: dict[str, int] = {}


class SessionResponse(BaseModel):
    id: str
    mode: str
    lang: str
    cpm: int
    accuracy: int
    duration: int
    sentences: int
    error_log: dict[str, int]
    created_at: datetime

    model_config = {"from_attributes": True}


class DailyStats(BaseModel):
    date: str   # 'YYYY-MM-DD'
    avg_cpm: float
    avg_accuracy: float
    session_count: int


class StatsResponse(BaseModel):
    daily: list[DailyStats]
    best_cpm: int
    total_sessions: int


class RankEntry(BaseModel):
    rank: int
    username: str
    best_cpm: int
    session_count: int
